vm/lib/src/io.rs

```diff
@@ -157,3 +156,28 @@ pub fn hint_slice(buf: &[u8]) {
   let mut my_reader = SyscallWriter { fd: FD_HINT };
   my_reader.write_all(buf).unwrap();
 }
+
+#[derive(Default)]
+pub struct Io {
+  read_remainder: std::collections::VecDeque<u8>,
+}
+
+impl std::io::Read for Io {
+  fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+    if self.read_remainder.is_empty() {
```

Comment:
I don't understand the use of `read_remainder` here, why is this necessary? In other words, why not just read directly using `io::read_vec()`? Probably worth an explanatory comment.

Reply:
Because `read()` is called with a small `buf` that doesn't fit the whole result of ``io::read_vec()`. The `remainder` buffer is needed to persist the rest between the calls. This is needed to support deserialization.

For example, say there is a struct which we want to serialize and write to `io` and read and deserialize on the other side. The struct is serialized to a single `Vec<u8>` and written to the io. This vec is later returned "whole" in a single `io::read_vec()`. But the deserializer like borsh might take multiple reads from the provided `reader` to fully deserialize back to the struct.

Reply:
makes sense. can we add a comment here to explain this? :pray: